core/src/main/java/hudson/model/queue/FutureImpl.java

```diff
@@ -95,4 +95,8 @@ public synchronized void setAsCancelled() {
     synchronized void addExecutor(@NonNull Executor executor) {
         this.executors.add(executor);
     }
+
+    synchronized void finished() {
+        executors.clear();
```

Comment:
https://github.com/jenkinsci/jenkins-test-harness/pull/643#discussion_r1304946075

Reply:
Yes https://github.com/jenkinsci/jenkins-test-harness/issues/672 is still on my list of things to look at, but I believe it is only tangentially related to this PR in that they both happen to involve code run from `WorkUnitContext.synchronizeEnd`; this PR is about the `executors` list, which is used only if `Future.cancel` is called, and does not interact directly with the `Future` done condition.